src/VisualStudio/Core/Def/Implementation/ProjectSystem/FileChangeWatcherProvider.cs

```diff
@@ -16,37 +17,32 @@ namespace Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem
-
         [ImportingConstructor]
         [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
-        public FileChangeWatcherProvider(IThreadingContext threadingContext, [Import(typeof(SVsServiceProvider))] Shell.IAsyncServiceProvider serviceProvider)
+        public FileChangeWatcherProvider(
+            IThreadingContext threadingContext,
+            IAsynchronousOperationListenerProvider listenerProvider,
+            [Import(typeof(SVsServiceProvider))] Shell.IAsyncServiceProvider serviceProvider)
         {
-            // We do not want background work to implicitly block on the availability of the SVsFileChangeEx to avoid any deadlock risk,
-            // since the first fetch for a file watcher might end up happening on the background.
-            Watcher = new FileChangeWatcher(_fileChangeService.Task);
-
-            System.Threading.Tasks.Task.Run(async () =>
+            var fileChangeService = Task.Factory.StartNew(
+                async () =>
                 {
-                    await threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync();
+                    await threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(threadingContext.DisposalToken);
 
                     var fileChangeService = (IVsAsyncFileChangeEx?)await serviceProvider.GetServiceAsync(typeof(SVsFileChangeEx)).ConfigureAwait(true);
                     Assumes.Present(fileChangeService);
-                    _fileChangeService.SetResult(fileChangeService);
-                });
+                    return fileChangeService;
+                },
+                threadingContext.DisposalToken,
+                TaskCreationOptions.RunContinuationsAsynchronously,
+                TaskScheduler.Default)
+                .Unwrap();
+
+            // We do not want background work to implicitly block on the availability of the SVsFileChangeEx to avoid any deadlock risk,
+            // since the first fetch for a file watcher might end up happening on the background.
+            Watcher = new FileChangeWatcher(listenerProvider, fileChangeService);
         }
 
         public FileChangeWatcher Watcher { get; }
-
-        // HACK HACK: this is to work around the SwitchToMainThread in the constructor above not
-        // being practical to run in unit tests. That SwitchToMainThread is working around a now-fixed
-        // bug in the shell where GetServiceAsync() might deadlock in the VS service manager
-        // if the UI thread was also dealing with the service at the same time. I'd remove the
-        // SwitchToMainThreadAsync right now instead of this doing this hack, but we're targeting this
-        // fix for a preview release that's too risky to do it in. Other options involve more extensive
-        // mocking or extracting of interfaces which is also just churn that will be immediately undone
-        // once we clean up the constructor either.
```

Comment:
This comment implies the SwitchToMainThread was a workaround, but we lost a comment explaining it was a workaround. Can we just delete it now? I'd have to find the internal conversation on this to understand it. My bad for not linking to something here.

Reply:
➡️ The comment was unnecessarily restrictive.

* The `SwitchToMainThreadAsync` is required for casting the object returned for `SVsFileChange`
* The workaround is no longer necessary for test code